### ISSUE TITLE
Managing unit files

### ISSUE THREAD

**Issue by web-refinery** [containers/podman#7813]
/kind feature

**Description**
Cockpit-podman is intended to do one thing and one thing well: be a frontend for podman. This raises an issue regarding autostarting pods (cockpit-project/cockpit-podman#364) which should be out-of-scope for cockpit but it might be in-scope for podman itself. 

In containers/podman#5494 @rhatdan asked whether there should be a --install flag for unit files and building on that, I'm wondering if podman should provide an interface (perhaps via a companion tool or a plugin) to manage the lifecycle of unit files (CRUD operations; update being `enable/disable`). This would simplify things for end users and software such as cockpit-podman.

(I did search for information in the issue tracker but I wasn't able to find much that was relevant, so apologies if I missed something and this was (partially) discussed and not taken into account when creating this issue.)

Cheers.

**Comment by vrothberg** [comment]
Thanks for reaching out, @web-refinery!

I share your thoughts, but for Podman. Having Podman manage the unit files can quickly turn into Podman acting as a front-end to `systemctl` which I think is outside of Podman's context. It's also hard to predict when the feature is "done" as we had to chase "systemctl".

A `podman generate systemd --install` flag, however, could make sense to at least erase the need for copying the files to the appropriate place for the current user.

In other words: anything beyond installing units should be done with `systemctl` (or dbus).

@baude @rhatdan WDYT?

**Comment by rhatdan** [comment]
Yes I think installing should be something we handle, especially where so many users trip over this.

**Comment by web-refinery** [comment]
Thanks for replying!

> I share your thoughts, but for Podman. Having Podman manage the unit files can quickly turn into Podman acting as a front-end to systemctl which I think is outside of Podman's context. It's also hard to predict when the feature is "done" as we had to chase "systemctl".

This makes sense.

> A podman generate systemd --install flag, however, could make sense to at least erase the need for copying the files to the appropriate place for the current user.

I would (if I hadn't known better) assume that after running this command that Podman would assume responsibility for the unit file and remove it whenever I tell it to remove the pod. 

> A `podman generate systemd --install` flag, however, could make sense to at least erase the need for copying the files to the appropriate place for the current user.

Adding a flag (i.e. `--systemd-unit`) to `podman-create` & `podman-rm` might be more intuitive. `podman-generate` wouldn't have to change.

**Comment by vrothberg** [comment]
Can you share a use case / your motivation to have Podman remove files?

So far, `podman generate systemd` generates units based on a container or pod. Running Podman in systemd services isn't straight forward but `generate systemd` makes this part easy. Managing the units is not something Podman is doing so far (only `podman auto-updates` restarts the units).

**Comment by web-refinery** [comment]
Cockpit-podman is a good example. It is intended to be a GUI for podman, nothing more and nothing less. So having podman create a unit file but not being able to remove it later on would leave dangling resources; obviously cockpit could handle removing the unit file itself but it wouldn't be able to do that (by default) when the pod was deleted in some other way (i.e. a user manually invoking podman-rm).

Just a thought; what about a small shell script to bridge the gap between podman & systemctl (not an entire interface) that only does 2 things: install & uninstall $POD's unit file. It can be used by users & by tools such as cockpit-podman. It would avoid the `--install` flag but it could use `podman-generate`.

**Comment by vrothberg** [comment]
> Cockpit-podman is a good example. It is intended to be a GUI for podman, nothing more and nothing less. So having podman create a unit file but not being able to remove it later on would leave dangling resources; obviously cockpit could handle removing the unit file itself but it wouldn't be able to do that (by default) when the pod was deleted in some other way (i.e. a user manually invoking podman-rm).

Does Cockpit-podman allow for generating unit files via podman-generate-systemd?

Note that most unit files have restart policies that will cause a pod/container to be recreated if it exited non-zero. Podman and systemd could hence run into a race of recreating a new container and deleting the unit file. We really want systemd to manage Podman (not the other way around) - `generate systemd` is meant to make that easier.

> Just a thought; what about a small shell script to bridge the gap between podman & systemctl (not an entire interface) that only does 2 things: install & uninstall $POD's unit file. It can be used by users & by tools such as cockpit-podman. It would avoid the --install flag but it could use podman-generate.

`podman generate systemd` is something I always expected to be part of some scripts but I am very hesitant to provide such a script upstream. We'd just defer the issue from Podman to a script but still had to maintain it. It's hard to meet the individual requirements so I prefer people to integrate `podman generate systemd` into their individual workflows.

Regarding installing/deinstalling the unit files:  that may be a nice feature request for `systemctl`.

**Comment by web-refinery** [comment]
> Does Cockpit-podman allow for generating unit files via podman-generate-systemd?

As far as I can tell, not yet, due to cockpit-podman wanting to only be a GUI for Podman. 

--

You've answered all my questions & filled in some blanks. I appreciate it, thanks!
